### ISSUE TITLE
[bug] Errors with Photon when changing scenes

### ISSUE THREAD

**Issue by timforever** [Placeholder-Software/Dissonance#25]
## Context
Using Photon for networking, when multiple players are connected and the scene is changed, multiple exceptions from Dissonance occur, and voice chat no longer functions.

## Expected Behavior
No exceptions should occur.

## Actual Behavior
The following exceptions occur, and voice chat stops working: 

```
DissonanceException: Error: Insufficient space in packet reader to read ushort! This is probably a bug in Dissonance, we're sorry! Please report the bug on the issue tracker "https://github.com/Placeholder-Software/Dissonance/issues". You could also seek help on the community at "http://placeholder-software.co.uk/dissonance/community" to get help for a temporary workaround. Error ID: 4AFBC61A-77D4-43B8-878F-796F0D921184
Dissonance.Networking.PacketReader.Check (Int32 count, System.String type) (at Assets/Plugins/Dissonance/Core/Networking/PacketReader.cs:57)
Dissonance.Networking.PacketReader.ReadUInt16 () (at Assets/Plugins/Dissonance/Core/Networking/PacketReader.cs:75)
Dissonance.Networking.BaseServer`3[Dissonance.Integrations.PhotonUnityNetworking.PhotonServer,Dissonance.Integrations.PhotonUnityNetworking.PhotonClient,System.Int32].NetworkReceivedPacket (Int32 source, ArraySegment`1 data) (at Assets/Plugins/Dissonance/Core/Networking/BaseServer.cs:121)
Dissonance.Integrations.PhotonUnityNetworking.PhotonServer.PacketDelivered (Byte eventcode, ArraySegment`1 data, Int32 senderid) (at Assets/Dissonance/Integrations/PhotonUnityNetworking/PhotonServer.cs:39)
Dissonance.Integrations.PhotonUnityNetworking.PhotonCommsNetwork.OnEvent (Byte eventcode, System.Object content, Int32 senderid) (at Assets/Dissonance/Integrations/PhotonUnityNetworking/PhotonCommsNetwork.cs:99)
NetworkingPeer.OnEvent (ExitGames.Client.Photon.EventData photonEvent) (at Assets/Photon Unity Networking/Plugins/PhotonNetwork/NetworkingPeer.cs:2573)
ExitGames.Client.Photon.PeerBase.DeserializeMessageAndCallback (System.Byte[] inBuff)
ExitGames.Client.Photon.EnetPeer.DispatchIncomingCommands ()
ExitGames.Client.Photon.PhotonPeer.DispatchIncomingCommands ()
PhotonHandler.Update () (at Assets/Photon Unity Networking/Plugins/PhotonNetwork/PhotonHandler.cs:157)
```

```
ArgumentNullException: Argument cannot be null.
Parameter name: data
Dissonance.Networking.Client.PacketRouter.NetworkReceivedPacket (ArraySegment`1 data) (at Assets/Plugins/Dissonance/Core/Networking/Client/PacketRouter.cs:54)
Dissonance.Networking.BaseClient`3[Dissonance.Integrations.PhotonUnityNetworking.PhotonServer,Dissonance.Integrations.PhotonUnityNetworking.PhotonClient,System.Int32].NetworkReceivedPacket (ArraySegment`1 data) (at Assets/Plugins/Dissonance/Core/Networking/BaseClient.cs:235)
Dissonance.Integrations.PhotonUnityNetworking.PhotonClient.PacketDelivered (Byte eventcode, ArraySegment`1 data, Int32 senderid) (at Assets/Dissonance/Integrations/PhotonUnityNetworking/PhotonClient.cs:40)
Dissonance.Integrations.PhotonUnityNetworking.PhotonCommsNetwork.OnEvent (Byte eventcode, System.Object content, Int32 senderid) (at Assets/Dissonance/Integrations/PhotonUnityNetworking/PhotonCommsNetwork.cs:99)
NetworkingPeer.OnEvent (ExitGames.Client.Photon.EventData photonEvent) (at Assets/Photon Unity Networking/Plugins/PhotonNetwork/NetworkingPeer.cs:2573)
ExitGames.Client.Photon.PeerBase.DeserializeMessageAndCallback (System.Byte[] inBuff)
ExitGames.Client.Photon.EnetPeer.DispatchIncomingCommands ()
ExitGames.Client.Photon.PhotonPeer.DispatchIncomingCommands ()
PhotonHandler.Update () (at Assets/Photon Unity Networking/Plugins/PhotonNetwork/PhotonHandler.cs:157)
```

## Steps to Reproduce
This can be reproduced using a lightly modified version of the Dissonance demo for Photon. It has simply been extended to load multiple scenes, which are just copies of each other. (By the way, I had to change a few things in the Matchmaking.cs code to get the demo to work at all. I suspect something in Photon has changed since the demo was created.)

1. Create a new Unity project.
2. Import PUN and Dissonance packages from store. Configure settings for Photon. Only include the Photon integration from Dissonance. 
3. Import the PhotonBug.unitypackage file. - [PhotonBug.zip](https://github.com/Placeholder-Software/Dissonance/files/920529/PhotonBug.zip)
4. Edit build settings so that it includes the 3 scenes located in Dissonance/Integrations/PhotonUnityNetworking/Demo. Enable "Development build".
5. Edit Player settings to enable running in background. Edit Input settings to enable push-to-talk.
5. Create a build so we can test multiplayer.
6. Run 2 instances of the build (or one in Unity) so they connect. Confirm that both instance are connected to each other.
7. The errors will occur when clicking the "Change scene" button on the instance running as the master client.

## Your Environment
- **Dissonance version used**: v1.0.8

- **Photon Unity Networking version used**: v1.83

- **Unity version**: 5.6.0f3

- **Editor Operating System and version**: Microsoft Windows 10 Pro 10.0.14393 (x64)

- **Build Settings**: Windows, x86_64, x86, Development Build

**Comment by martindevans** [comment]
Thanks for the report Tim, that's definitely not one we've seen before! Both of these failures seem to be versions of the same thing - Photon gives us an invalid packet (the first one has fails to read the very first 2 bytes from the packet, in the second the packet is simply null). I'll check out your reproduction project to see if I get the same thing.

**Comment by martindevans** [comment]
I can reproduce this, thanks for the perfect reproduction project :)

**Comment by timforever** [comment]
Thanks Martin!

You may have noticed - I got tired of working with UNET and am giving Photon a try. 😆

Also, thank you for getting the spatialization issue fixed!

**Comment by martindevans** [comment]
Hi Tim,

I looked into this problem a little more today and it looks like the problem is related to the fact that each new scene tears down a Dissonance component and sets up a new one. I'm not certain why this causes the precise issues you're getting but it doesn't surprise me things get confused.

We have a large internal feature for improving the lifetime management of Dissonance components which will solve a few related issues (yours don't live long enough, #23 live for too long).

**Comment by timforever** [comment]
Thanks Martin. Any idea of when this improvement might be available?

**Comment by martindevans** [comment]
I should be wrapping up my current work today or tomorrow and then I'm planning to move onto this. I'm expecting it to be a fairly large change in how lifetimes are managed so it could take a while to get right.

The general plan is to apply `Object.DontDestroyOnLoad` to the main Dissonance entity (the one with DissonanceComms and your network integration attached). Then you can create a DissonanceComms component as early as you like and the component should gracefully handle living across multiple network sessions. I haven't investigated the practicalities of this yet though, so that might all change!

**Comment by martindevans** [comment]
Ok I've got a fix for this bug sooner than I expected! I had thought that the problem was caused by there being two Dissonance instances both trying to interact with the same network session and confusing each other. It turns out that was wrong, this is actually a bug in the shutdown logic for the photon integration - it doesn't exist!

Add this in to PhotonCommsNetwork.cs:

    protected override void Shutdown()
    {
        base.Shutdown();
        PhotonNetwork.OnEventCall -= OnEvent;
    }

Adding this seems to completely eliminate the problem in your reproduction project. I should mention I've only tested this using the latest development branch of Dissonance but I don't expect that to be a problem.

**Comment by timforever** [comment]
Thanks so much Martin, this is fixing the issue in my project as well. It's working great now.

**Comment by xafgun** [comment]
Yes, I solved problem, too.

**Comment by martindevans** [comment]
Oops this was auto closed when the PR was merged in the dev branch. I'll re-open this and mark it as awaiting merge.

**Comment by martindevans** [comment]
This was released in [Dissonance 1.0.9](https://placeholder-software.co.uk/dissonance/releases/1.0.9.html) which is currently live on the asset store. As such I shall close this issue - please feel free to reopen it if it's still a problem :)
